common/cmdparse: use string_view for the key and return val by retval

tchaikov opened the pull request:
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

rzarzynski commented on src/common/cmdparse.h:
```diff
@@ -96,9 +97,9 @@ bool cmd_getval(
+cmd_putval(CephContext *cct, cmdmap_t& cmdmap, std::string_view k, const T& val)
 {
-  cmdmap[k] = val;
+  cmdmap.insert_or_assign(std::string{k}, val);
```
This might impose some extra work, I guess. Some `std::string` implementations use shared buffer techniques (COW) to deduplicate the memory while the `std::string` constructor, IIUC, copies `k.size()` bytes from `k.data()` into a fresh region.

However, I don't expect `cmd_putval()` nor even `cmd_getval()` to be on hot paths, and thus the performance shouldn't be really a concern.

rzarzynski commented on src/common/cmdparse.h:
```diff
@@ -78,19 +78,16 @@ bool cmd_getval(const cmdmap_t& cmdmap,
-  const cmdmap_t& cmdmap, std::string_view k,
-  T& val, const T& defval)
+template <typename T, typename V>
+T cmd_getval(const cmdmap_t& cmdmap, std::string_view k,
```
:+1:

tchaikov commented on src/common/cmdparse.cc:
```diff
@@ -651,7 +651,8 @@ bool cmd_getval(const cmdmap_t& cmdmap,
       try {
-        std::string expected = "--" + k;
+        std::string expected{"--"};
+	expected += k;
```
fixed already.

tchaikov commented:
jenkins test api

tchaikov commented:
changelog:

- add a commit to link rgw_a against fmt::fmt. to address following failure
```
[ 95%] Linking CXX executable ../../bin/radosgw-admin
../../lib/librgw_a.a(rgw_lc.cc.o): In function `rgw::lc::s3_expiration_header(DoutPrefixProvider*, rgw_obj_key const&, RGWObjTags const&, std::chrono::time_point<ceph::real_clock, std::chrono::duration<unsigned long, std::ratio<1l, 1000000000l> > > const&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, ceph::buffer::v15_2_0::list, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> > > const&)':
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-4428-g1fbb5314/rpm/el8/BUILD/ceph-17.0.0-4428-g1fbb5314/src/fmt/include/fmt/format.h:3490: undefined reference to `fmt::v6::buffer_context<std::back_insert_iterator<fmt::v6::detail::buffer<char> >, char>::iterator fmt::v6::detail::vformat_to<char>(fmt::v6::detail::buffer<char>&, fmt::v6::basic_string_view<char>, fmt::v6::basic_format_args<fmt::v6::buffer_context<std::back_insert_iterator<fmt::v6::detail::buffer<fmt::v6::type_identity<char>::type> >, fmt::v6::type_identity<char>::type> >)'
../../lib/librgw_a.a(rgw_datalog.cc.o): In function `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > fmt::v6::detail::vformat<char>(fmt::v6::basic_string_view<char>, fmt::v6::basic_format_args<fmt::v6::basic_format_context<std::back_insert_iterator<fmt::v6::detail::buffer<fmt::v6::type_identity<char>::type> >, fmt::v6::type_identity<char>::type> >)':
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-4428-g1fbb5314/rpm/el8/BUILD/ceph-17.0.0-4428-g1fbb5314/src/fmt/include/fmt/format.h:3490: undefined reference to `fmt::v6::buffer_context<std::back_insert_iterator<fmt::v6::detail::buffer<char> >, char>::iterator fmt::v6::detail::vformat_to<char>(fmt::v6::detail::buffer<char>&, fmt::v6::basic_string_view<char>, fmt::v6::basic_format_args<fmt::v6::buffer_context<std::back_insert_iterator<fmt::v6::detail::buffer<fmt::v6::type_identity<char>::type> >, fmt::v6::type_identity<char>::type> >)'
../../lib/librgw_a.a(rgw_auth_s3.cc.o): In function `rgw::auth::s3::gen_v4_scope(std::chrono::time_point<ceph::real_clock, std::chrono::duration<unsigned long, std::ratio<1l, 1000000000l> > > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-4428-g1fbb5314/rpm/el8/BUILD/ceph-17.0.0-4428-g1fbb5314/src/fmt/include/fmt/format.h:3490: undefined reference to `fmt::v6::buffer_context<std::back_insert_iterator<fmt::v6::detail::buffer<char> >, char>::iterator fmt::v6::detail::vformat_to<char>(fmt::v6::detail::buffer<char>&, fmt::v6::basic_string_view<char>, fmt::v6::basic_format_args<fmt::v6::buffer_context<std::back_insert_iterator<fmt::v6::detail::buffer<fmt::v6::type_identity<char>::type> >, fmt::v6::type_identity<char>::type> >)'
collect2: error: ld returned 1 exit status

```

tchaikov commented:
jenkins test api

tchaikov commented:
jenkins test api

tchaikov commented:
jenkins test make check

batrick reviewed (changes requested):
Looks like this PR causes a number of failures in the fs suite. One concrete example from

https://pulpito.ceph.com/pdonnell-2021-05-20_21:51:48-fs-wip-pdonnell-testing-20210520.194920-distro-basic-smithi/

is

````
2021-05-20T22:47:07.469+0000 7fddc17fa700  1 -- [v2:172.21.15.50:3300/0,v1:172.21.15.50:6789/0] --> v1:172.21.15.50:6819/2728367922 -- mon_command_ack([{"prefix":"osd blocklist", "blocklistop":"add","addr":"v1:172.21.15.200:0/4263861803"}]=0 blocklisting 172.21.15.200:0/4263861803 until 2021-05-20T22:47:06.468840+0000 (0 sec) v36) v1 -- 0x7fdd88089c20 con 0x7fddbc11e330
````

From: /ceph/teuthology-archive/pdonnell-2021-05-20_21:51:48-fs-wip-pdonnell-testing-20210520.194920-distro-basic-smithi/6125435/remote/smithi050/log/ceph-mon.a.log.gz

default blocklist timeout from the config is 0 instead of 1 hour.

tchaikov commented:
@batrick thank you for testing. just fixed this issue. i *moved* the default value away by passing a rvalue reference of the default value and perfect forwarding it in the new implementation of `cmd_getval()` which returns `T` instead of `bool`. should just pass it by const reference.

@rzarzynski i fixed the issue above, could you please take another look?

tchaikov commented on src/common/cmdparse.h:
```diff
@@ -79,19 +79,16 @@ bool cmd_getval(const cmdmap_t& cmdmap,
   if (found == cmdmap.end()) {
-    val = defval;
-    return true;
+    return T(defval);
```
i fixed by returning `T(defval)`, instead of `std::forward<V>(defval)`.

rzarzynski replied:
Huh, do we really had / still have an usage that passed rvalue reference and assumed the callee won't move-out from it? If so, this might be an indicator of another problem.

tchaikov replied:
you are right, we don't do this actually. but after a second thought. i think we should not use perfect forwarding here. for two reasons

- we never call this function like `cmd_getval_or(..., std::move(foo))`, so unless we have `cmd_getval_or(..., V& defval)`, we are not allowed to move from `defval`.
- we cannot benefit from constructing the return value by perfect forwarding an argument. unless we have something like
  ```c++
  string val = cmd_getval_or<string>(cmdmap, "val_name", std::string("foo"));
  ```
  but what we have in general is
  ```c++
  string val = cmd_getval_or<string>(cmdmap, "val_name", "foo");
  ```

tchaikov commented on src/common/cmdparse.h:
```diff
@@ -79,19 +79,16 @@ bool cmd_getval(const cmdmap_t& cmdmap,
-  const cmdmap_t& cmdmap, std::string_view k,
-  T& val, const T& defval)
+template <typename T, typename V>
+T cmd_getval_or(const cmdmap_t& cmdmap, std::string_view k,
```
also, changed the function signature so it's different from the one defined at line 65. otherwise, that one would be chosen instead of this one when the compiler is performing argument-dependent lookup.

rzarzynski commented on src/mgr/DaemonServer.cc:
```diff
@@ -1408,8 +1408,7 @@ bool DaemonServer::_handle_command(
       prefix == "osd test-reweight-by-utilization";
-    int64_t oload;
-    cmd_getval(cmdctx->cmdmap, "oload", oload, int64_t(120));
+    int64_t oload = cmd_getval_or<int64_t>(cmdctx->cmdmap, "oload", 120);
```
The new naming seems to be more explicit.

rzarzynski commented:
@tchaikov: I don't see any obvious problem in the revised changeset. I'm curious about the usage that passed rvalue.

tchaikov commented:
- https://pulpito.ceph.com/kchai-2021-05-21_12:49:52-rados-wip-kefu-testing-2021-05-21-1851-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-22_15:42:09-rados-wip-kefu-testing-2021-05-22-2207-distro-basic-smithi/

batrick commented:
https://pulpito.ceph.com/pdonnell-2021-05-23_00:57:16-fs-wip-kefu-testing-2021-05-22-2207-distro-basic-smithi/

(I'm getting a bunch of 503 errors when scheduling so this might not be the full set I want.)

tchaikov commented:
jenkins test api

tchaikov commented:
> https://pulpito.ceph.com/pdonnell-2021-05-23_00:57:16-fs-wip-kefu-testing-2021-05-22-2207-distro-basic-smithi/
> 
> (I'm getting a bunch of 503 errors when scheduling so this might not be the full set I want.)

@batrick thank you for rerunning the tests with my branch. i am rerunning the failed test in your batch at https://pulpito.ceph.com/kchai-2021-05-23_02:53:03-fs-wip-pdonnell-testing-20210520.194920-updated-distro-basic-smithi/ by 

1. reverting the bugging patchset from your branch at ceph-ci and 
2. applying the latest revision
3. building the updated branch
4. rerunning the failed run.

hopefully this will do the trick.

tchaikov commented:
@batrick i suspect there are some offending commits which break the test. i reran the failed tests after reapplying my latest change, but all tests either failed or have been running for hours before being killed. see https://pulpito.ceph.com/kchai-2021-05-23_12:55:24-fs-wip-pdonnell-testing-20210520.194920-updated-distro-basic-smithi/

but if i just rerun this PR applied on top of master against those tests, i have https://pulpito.ceph.com/kchai-2021-05-23_14:34:44-fs-wip-cmd-getval-kefu-distro-basic-smithi/

tchaikov commented:
@batrick ping?

tchaikov commented:
@batrick i dismissed your nack. as i've rerun the failed tests. most of them passed.

batrick commented:
No worries. I'm on PTO so sorry about no response.
